src/main/kotlin/org/partiql/scribe/sql/RexConverter.kt

```diff
@@ -1,448 +1,595 @@
+                PType.REAL -> DataType.REAL()
+                PType.DOUBLE -> DataType.DOUBLE_PRECISION()
+                // String types
+                PType.CHAR -> DataType.CHAR()
+                PType.VARCHAR -> DataType.VARCHAR()
+                PType.STRING -> DataType.STRING()
+                // Dynamic type
+                PType.DYNAMIC -> return value // TODO possibly move into separate rewrite
+                PType.BAG -> return value
+                else ->
+                    listener.reportAndThrow(
+                        ScribeProblem.simpleError(
+                            code = ScribeProblem.UNSUPPORTED_PLAN_TO_AST_CONVERSION,
+                            message = "CAST with type ${rex.target.code()} is not yet supported",
+                        ),
+                    )
+            }
+        return exprCast(
+            value = value,
+            asType = targetType,
+        )
     }
 
-    override fun visitRexOpPathKey(node: Rex.Op.Path.Key, ctx: StaticType): Expr {
-        val prev = visitRex(node.root, ctx)
-        val step = exprPathStepIndex(visitRex(node.key, ctx))
-        return if (prev is Expr.Path) {
-            exprPath(prev.root, prev.steps + step)
-        } else {
-            exprPath(visitRex(node.root, ctx), listOf(step))
-        }
+    override fun visitCoalesce(
+        rex: RexCoalesce,
+        ctx: Unit,
+    ): Expr {
+        val args = rex.args.map { visitRex(it, ctx) }
+        return exprCoalesce(args)
     }
 
-    override fun visitRexOpPathSymbol(node: Rex.Op.Path.Symbol, ctx: StaticType): Expr {
-        val prev = visitRex(node.root, ctx)
-        val step = exprPathStepSymbol(id(node.key))
-        return if (prev is Expr.Path) {
-            exprPath(prev.root, prev.steps + step)
-        } else {
-            exprPath(visitRex(node.root, ctx), listOf(step))
-        }
+    override fun visitDispatch(
+        rex: RexDispatch,
+        ctx: Unit,
+    ): Expr {
+        val fn = rex.functions.first()
+        val args = rex.args.map { SqlArg(visitRex(it, ctx), it.type.pType) }
+        return transform.getFunction(fn.signature.name, args)
     }
 
-    override fun visitRexOpCollection(node: Rex.Op.Collection, ctx: StaticType): Expr {
-        val values = node.values.map { visitRex(it, ctx) }
-        return when (ctx) {
-            is ListType -> Expr.Collection(Expr.Collection.Type.LIST, values)
-            is SexpType -> Expr.Collection(Expr.Collection.Type.SEXP, values)
-            is BagType -> Expr.Collection(Expr.Collection.Type.BAG, values)
-            else -> throw UnsupportedOperationException("unsupported collection type $ctx")
-        }
+    override fun visitError(
+        rex: RexError,
+        ctx: Unit,
+    ): Expr {
+        listener.reportAndThrow(
+            ScribeProblem.simpleError(
+                code = ScribeProblem.UNSUPPORTED_PLAN_TO_AST_CONVERSION,
+                message = "ERROR is not yet supported",
```

Comment:
An error node at this point means a parse that failed in some way? Should we have a more informative error message?

Reply:
More specifically, an error here means a query successfully parsed to the AST, but there's some error that occurred when creating the typed plan. The error listener for PLK, `PErrorListener`, should have already reported the error so this node should not be visited.